uaclient/event_logger.py

```diff
@@ -0,0 +1,135 @@
+
+    if _event_logger is None:
+        _event_logger = EventLogger()
+
+    return _event_logger
+
+
+@enum.unique
+class EventLoggerMode(enum.Enum):
+    """
+    Defines event logger supported modes.
+    Currently, we only support the cli and machine-readable
+    mode. On cli mode, we will print to stdout/stderr any
+    event that we receive. On machine-readable mode, we will
+    store those events and parse them for an specified
+    format.
+    """
+
+    CLI = object()
+    MACHINE_READABLE = object()
+
+
+class EventLogger:
+    def __init__(self):
+        self._error_events = []  # type: List[Dict[str, Optional[str]]]
+        self._warning_events = []  # type: List[Dict[str, Optional[str]]]
+        self._processed_services = []  # type: List[str]
+        self._failed_services = set()  # type: Set[str]
+        self._needs_reboot = False
+
+        # By default, the event logger will be on CLI mode,
+        # printing every event it receives.
+        self._event_logger_mode = EventLoggerMode.CLI
+
+    def reset(self):
+        self._error_events = []
+        self._warning_events = []
+        self._processed_services = []
+        self._failed_services = set()
+        self._needs_reboot = False
+
+    def set_event_mode(self, event_mode: EventLoggerMode):
+        self._event_logger_mode = event_mode
+
+    def info(self, info_msg: str):
+        if self._event_logger_mode == EventLoggerMode.CLI:
+            print(info_msg)
+
+    def _record_dict_event(
+        self,
+        msg: str,
+        service: Optional[str],
+        event_dict: List[Dict[str, Optional[str]]],
+    ):
+        event_type = "service" if service else "system"
+        event_dict.append(
+            {"type": event_type, "service": service, "message": msg}
+        )
+
+    def error(self, error_msg: str, service: Optional[str] = None):
+        if self._event_logger_mode == EventLoggerMode.MACHINE_READABLE:
+            self._record_dict_event(
+                msg=error_msg, service=service, event_dict=self._error_events
+            )
+
+    def warning(self, warning_msg: str, service: Optional[str] = None):
+        if self._event_logger_mode == EventLoggerMode.MACHINE_READABLE:
+            self._record_dict_event(
+                msg=warning_msg,
+                service=service,
+                event_dict=self._warning_events,
+            )
+
+    def raise_error(self, exception: Exception, service: Optional[str] = None):
+        if self._event_logger_mode == EventLoggerMode.CLI:
+            raise exception
+        else:
+            error_msg = getattr(exception, "msg") or str(exception)
+            self.error(error_msg=error_msg, service=service)
+
+    def raise_error_and_finish(
+        self, exception: Exception, service: Optional[str] = None
+    ):
+        self.raise_error(exception=exception, service=service)
+
+        if self._event_logger_mode == EventLoggerMode.MACHINE_READABLE:
+            print(self.process_events())
```

Comment:
I wouldn't expect this module to be the one calling `print`. in this case, wouldn't this just return a dict or json.dumps string and the caller would call `print`? (This doesn't matter a whole lot, but makes the boundaries clearer for testing IMO).

Reply:
Originally, that was my intention, but the problem is that if we do that, we will loose the "real-time" aspect of the cli mode. We would print all the messages at the same time, which can be bad if the user is running a slow command, like enabling fips

Reply:
Right, my comment was more talking about the json output at the end though. But per my comment in the other thread, it is necessary for this print to be in this module as well because it should only happen in machine_readable mode.